service/operator/service_test.go

```diff
@@ -1,95 +1,604 @@
+						{
+							IngressPort: 30010,
+							Protocol:    "http",
+							LBPort:      31000,
+						},
+					},
 				},
-				{
-					Name: "test-port-two",
-					Port: 30002,
+			},
+			Expected: DesiredState{
+				ConfigMapData: map[string]string{
+					"31000": "al9qy/worker:30010",
+				},
+				ServicePorts: []apiv1.ServicePort{
+					{
+						Name:       "http-30010-al9qy",
+						Protocol:   apiv1.ProtocolTCP,
+						Port:       int32(31000),
+						TargetPort: intstr.FromInt(31000),
+						NodePort:   int32(31000),
+					},
+				},
+			},
+			ErrorMatcher: nil,
+		},
+		{
+			Obj: &ingresstpr.CustomObject{
+				Spec: ingresstpr.Spec{
+					GuestCluster: guestcluster.GuestCluster{
+						ID:        "p1l6x",
+						Namespace: "p1l6x",
+						Service:   "worker",
+					},
+					HostCluster: hostcluster.HostCluster{
+						IngressController: ingresscontroller.IngressController{
+							ConfigMap: "ingress-controller",
+							Namespace: "kube-system",
+							Service:   "ingress-controller",
+						},
+					},
+					ProtocolPorts: []protocolport.ProtocolPort{
+						{
+							IngressPort: 30010,
+							Protocol:    "http",
+							LBPort:      31000,
+						},
+						{
+							IngressPort: 30011,
+							Protocol:    "https",
+							LBPort:      31001,
+						},
+						{
+							IngressPort: 30012,
+							Protocol:    "udp",
+							LBPort:      31002,
+						},
+					},
+				},
+			},
+			Expected: DesiredState{
+				ConfigMapData: map[string]string{
+					"31000": "p1l6x/worker:30010",
+					"31001": "p1l6x/worker:30011",
+					"31002": "p1l6x/worker:30012",
+				},
+				ServicePorts: []apiv1.ServicePort{
+					{
+						Name:       "http-30010-p1l6x",
+						Protocol:   apiv1.ProtocolTCP,
+						Port:       int32(31000),
+						TargetPort: intstr.FromInt(31000),
+						NodePort:   int32(31000),
+					},
+					{
+						Name:       "https-30011-p1l6x",
+						Protocol:   apiv1.ProtocolTCP,
+						Port:       int32(31001),
+						TargetPort: intstr.FromInt(31001),
+						NodePort:   int32(31001),
+					},
+					{
+						Name:       "udp-30012-p1l6x",
+						Protocol:   apiv1.ProtocolTCP,
+						Port:       int32(31002),
+						TargetPort: intstr.FromInt(31002),
+						NodePort:   int32(31002),
+					},
+				},
+			},
+			ErrorMatcher: nil,
+		},
+	}
+
+	var err error
+	var newService *Service
+	{
+		newConfig := DefaultConfig()
+		newService, err = New(newConfig)
+		if err != nil {
+			t.Fatal("expected", nil, "got", err)
+		}
+	}
+
+	for i, testCase := range testCases {
+		result, err := newService.GetDesiredState(testCase.Obj)
+		if err != nil && testCase.ErrorMatcher == nil {
+			t.Fatal("case", i+1, "expected", nil, "got", err)
+		}
+		if testCase.ErrorMatcher != nil && !testCase.ErrorMatcher(err) {
+			t.Fatal("case", i+1, "expected", true, "got", false)
+		}
+		e, ok := result.(DesiredState)
+		if !ok {
+			t.Fatalf("case %d expected %#v got %#v", i+1, true, false)
+		}
+		if !reflect.DeepEqual(testCase.Expected, e) {
```

Comment:
What about `assert` from https://github.com/stretchr/testify? Code will be shorter and we're already using that lib in the other projects.